presto-spi/src/main/java/com/facebook/presto/spi/predicate/TupleDomain.java

```diff
@@ -119,6 +120,27 @@ private TupleDomain(Optional<Map<T, Domain>> domains)
                 .collect(toLinkedMap(Map.Entry::getKey, entry -> new ConstantExpression(entry.getValue().getNullableSingleValue(), entry.getValue().getType()))));
     }
 
+    /**
+     * Extract all column constraints that require a set of one (or more) values in their respective Domains. All the
+     * ranges (see {@link SortedRangeSet}) must have fixed values.
+     * Returns an empty Optional if the Domain is none.
+     */
+    public static <T> Optional<Map<T, Set<NullableValue>>> extractFixedValuesFromRanges(TupleDomain<T> tupleDomain)
+    {
+        if (!tupleDomain.getDomains().isPresent()) {
+            return Optional.empty();
+        }
+
+        return Optional.of(tupleDomain.getDomains().get().entrySet().stream()
+                .filter(entry -> entry.getValue().getValues().getType().isOrderable())
```

Comment:
@gparai this is a bit difficult to read and it seems like a lot of computation is repeated unnecessarily; consider rewriting for clarity and efficiency:

```
    public static <T> Optional<Map<T, Set<NullableValue>>> extractFixedValuesFromRanges(TupleDomain<T> tupleDomain)
    {
        if (!tupleDomain.getDomains().isPresent()) {
            return Optional.empty();
        }

        Map<T, Set<NullableValue>> fixedValues = new HashMap<>();
        for (Map.Entry<T, Domain> entry : tupleDomain.getDomains().get().entrySet()) {
            Optional<Set<NullableValue>> values = extractFixedValues(entry.getValue());
            if (values.isPresent()) {
                fixedValues.put(entry.getKey(), values.get());
            }
        }

        return Optional.of(fixedValues);
    }

    private static Optional<Set<NullableValue>> extractFixedValues(Domain domain)
    {
        ValueSet values = domain.getValues();
        if (!values.getType().isOrderable()) {
            return Optional.empty();
        }

        List<Range> ranges = values.getRanges().getOrderedRanges();
        boolean allSingleValue = ranges.stream().allMatch(Range::isSingleValue);
        if (!allSingleValue) {
            return Optional.empty();
        }

        return Optional.of(ranges.stream()
                .map(range -> new NullableValue(domain.getType(), range.getSingleValue()))
                .collect(Collectors.toSet()));
    }
```

Reply:
Done